src/Compilers/CSharp/Portable/Lowering/LambdaRewriter/LambdaRewriter.Analysis.Tree.cs

```diff
@@ -0,0 +1,449 @@
+                        {
+                            visitStack.Push(closure.Symbol);
+                        }
+                    }
+                });
+
+                while (visitStack.Count > 0)
+                {
+                    var current = visitStack.Pop();
+                    var setToAddTo = capturesVariable.Contains(current) ? capturesVariable : capturesThis;
+                    foreach (var capturesCurrent in methodGraph[current])
+                    {
+                        if (setToAddTo.Add(capturesCurrent))
+                        {
+                            visitStack.Push(capturesCurrent);
+                        }
+                    }
+                }
+
+                VisitClosures(_scopeTree, (scope, closure) =>
+                {
+                    if (!capturesVariable.Contains(closure.Symbol))
+                    {
+                        closure.CapturedVariables.Clear();
+                    }
+                    if (capturesThis.Contains(closure.Symbol))
+                    {
+                        closure.CapturedVariables.Add(_topLevelMethod.ThisParameter);
+                    }
+                });
+            }
+
+            /// <summary>
+            /// Visit all closures in all nested scopes and run the <paramref name="action"/>.
+            /// </summary>
+            private static void VisitClosures(Scope scope, Action<Scope, Closure> action)
+            {
+                foreach (var closure in scope.Closures)
+                {
+                    action(scope, closure);
+                }
+                foreach (var nested in scope.NestedScopes)
+                {
+                    VisitClosures(nested, action);
+                }
+            }
+
+            private class ScopeTreeBuilder : BoundTreeWalkerWithStackGuardWithoutRecursionOnTheLeftOfBinaryOperator
+            {
+                private Scope _currentScope;
+                // Null if we're not inside a closure scope, otherwise the nearest closure scope
+                private Closure _currentClosure;
+                private bool _inExpressionTree = false;
+
+                // A mapping from all captured vars to the scope they were declared in. This
+                // is used when recording captured variables as we must know what the lifetime
+                // of a captured variable is to determine the lifetime of its capture environment.
+                private readonly SmallDictionary<Symbol, Scope> _localToScope = new SmallDictionary<Symbol, Scope>();
+
+                private readonly Analysis _analysis;
+
+                private ScopeTreeBuilder(Scope rootScope, Analysis analysis)
+                {
+                    _currentScope = rootScope;
+                    _analysis = analysis;
+                    _currentClosure = null;
+                }
+
+                public static Scope Build(BoundNode node, Analysis analysis)
```

Comment:
Is it legal to call `Build` with a `BoundNode` for say a `foreach` loop? Or is it only valid for top level nodes? If only top level nodes is there an easy assertion we can make here?